File: agrismart front end/backend/irrigation.py
import pandas as pd
import requests

# Tomorrow.io API details
API_KEY = 'jiHcUKYSrkFxXEclxkxKQy0FkhuVHP1r'  # Replace with your valid API key

# Data for Monthly Irrigation Cycles of Various Crops
data = {
    "Crop": ["Rice", "Rice", "Rice", "Rice", "Rice", "Rice",
             "Maize", "Maize", "Maize", "Maize", "Maize", "Maize",
             "Groundnut", "Groundnut", "Groundnut", "Groundnut", "Groundnut", "Groundnut",
             "Red Gram", "Red Gram", "Red Gram", "Red Gram", "Red Gram", "Red Gram"],
    "Month": ["June", "July", "August", "September", "October", "November"] * 4,
    "Growth Stage": [
        "Land preparation, Sowing (Nursery)", "Transplanting, Early Vegetative", 
        "Vegetative (Tillering)", "Reproductive (Panicle initiation)", 
        "Flowering and Grain Filling", "Maturity and Harvest",
        "Sowing", "Vegetative Growth", "Tasseling", "Silking", 
        "Grain Filling", "Maturity",
        "Land Preparation and Sowing", "Vegetative Growth", 
        "Flowering", "Pod Development", "Seed Filling", "Maturity",
        "Sowing", "Vegetative Growth", "Branching", "Flowering", 
        "Pod Development", "Maturity"
    ],
    "Irrigation Depth per Application (mm)": [50, 50, 75, 80, 80, 50,
                                              40, 50, 60, 70, 70, 50,
                                              30, 40, 50, 55, 60, 40,
                                              30, 40, 50, 60, 60, 50],
    "Frequency (Days)": ["2-3", "3-4", "4-5", "5-6", "7", "As needed (if dry)"] * 4,
    "Water Requirement (mm/month)": [300, 250, 400, 500, 400, 100,
                                     150, 200, 300, 350, 300, 100,
                                     120, 180, 250, 300, 280, 80,
                                     100, 150, 220, 250, 300, 120]
}

df = pd.DataFrame(data)

# Function to get weather data and recommendations
def get_weather_and_recommendations(lat, lon, crop, crop_stage, soil_moisture):
    url = f"https://api.tomorrow.io/v4/timelines?location={lat},{lon}&fields=temperature,humidity,windSpeed,precipitationIntensity&timesteps=1d&timezone=auto&apikey={API_KEY}"

    response = requests.get(url)

    if response.status_code == 200:
        data = response.json()
        intervals = data['data']['timelines'][0]['intervals']

        # Extract weather data
        rain_forecast = False
        rain_dates = []
        total_rainfall = 0

        # Check rainfall forecast for the next 4-5 days
        for i, day in enumerate(intervals[:5]):
            date = day['startTime'][:10]
            rain_amount = day['values'].get('precipitationIntensity', 0)
            total_rainfall += rain_amount

            if rain_amount > 0:
                rain_forecast = True
                rain_dates.append(date)

        # Weather output
        print("\n--- Current Weather ---")
        today = intervals[0]['values']
        print(f"Temperature: {today['temperature']}°C")
        print(f"Humidity: {today['humidity']}%")
        print(f"Wind Speed: {today['windSpeed']} m/s")

        # Irrigation timing recommendations
        if total_rainfall >= 10:
            print(f"\nIrrigation Timing: Rain is expected over the next few days ({', '.join(rain_dates)}).")
            print("Skip irrigation to avoid over-watering.")
        else:
            print("\nIrrigation Timing: Dry weather is expected over the next 4-5 days.")
            if soil_moisture < 40:
                print("Soil moisture is low, irrigation is needed within the next 24 hours.")
            else:
                print("Soil moisture is adequate; monitor for changes over the next few days.")

        # Water quantity recommendations
        water_quantity = get_water_quantity(crop_stage, soil_moisture)
        print(f"\nWater Quantity: For your current crop stage, apply {water_quantity} mm of water to the field.")

        # Frequency recommendations
        frequency = df.loc[(df['Crop'] == crop) & (df['Growth Stage'] == crop_stage), 'Frequency (Days)'].values[0]
        print(f"\nFrequency: Irrigate every {frequency} days to maintain optimal moisture.")

        # Irrigation efficiency tips
        print("\nIrrigation Efficiency Tips:")
        print("1. Use drip irrigation or furrow irrigation to minimize water loss.")
        print("2. Water early in the morning or late in the evening to reduce evaporation.")

    else:
        print(f"Error: {response.status_code}, {response.text}")

# Helper function to determine water quantity based on crop stage and soil moisture
def get_water_quantity(stage, soil_moisture):
    base_water = df.loc[df['Growth Stage'] == stage, 'Irrigation Depth per Application (mm)'].values[0]

    # Adjust water quantity based on soil moisture levels
    if soil_moisture < 30:
        return base_water + 10  # Increase by 10mm if soil is very dry
    elif 30 <= soil_moisture < 50:
        return base_water  # Standard water requirement
    else:
        return max(base_water - 10, 10)  # Reduce water if soil is sufficiently moist

# Function to display irrigation cycle for the selected crop
def display_irrigation_cycle(crop):
    print(f"\n--- Monthly Irrigation Cycle for {crop} ---")
    crop_data = df[df['Crop'] == crop]
    print(crop_data.to_string(index=False))

# Example usage
def main():
    lat = input("Enter Latitude: ")
    lon = input("Enter Longitude: ")
    crop = input("Enter Crop (Rice, Maize, Groundnut, Red Gram): ").capitalize()

    if crop not in df['Crop'].unique():
        print("Invalid crop selection.")
        return

    print("\nSelect the Current Growth Stage:")
    stages = df[df['Crop'] == crop]['Growth Stage'].unique()
    for i, stage in enumerate(stages):
        print(f"{i + 1}. {stage}")
    stage_index = int(input("Enter the stage number: ")) - 1
    crop_stage = stages[stage_index]

    soil_moisture = int(input("Enter current soil moisture level (0-100): "))

    display_irrigation_cycle(crop)
    get_weather_and_recommendations(lat, lon, crop, crop_stage, soil_moisture)

# Run the main function
main()

File: agrismart front end/irrigation.js
// Crop Data for Monthly Irrigation Cycles
const cropData = [
    { crop: "Rice", month: "June", stage: "Land preparation, Sowing (Nursery)", depth: 50, frequency: "2-3", waterRequirement: 300 },
    { crop: "Rice", month: "July", stage: "Transplanting, Early Vegetative", depth: 50, frequency: "3-4", waterRequirement: 250 },
    { crop: "Rice", month: "August", stage: "Vegetative (Tillering)", depth: 75, frequency: "4-5", waterRequirement: 400 },
    { crop: "Maize", month: "July", stage: "Vegetative Growth", depth: 50, frequency: "3-4", waterRequirement: 200 },
    { crop: "Maize", month: "August", stage: "Tasseling", depth: 60, frequency: "4-5", waterRequirement: 300 },
    { crop: "Groundnut", month: "June", stage: "Land Preparation and Sowing", depth: 30, frequency: "2-3", waterRequirement: 120 },
    { crop: "Red Gram", month: "August", stage: "Branching", depth: 50, frequency: "4-5", waterRequirement: 220 }
];

// Tomorrow.io API details
const API_KEY = 'PtOfldhuA5hMqeeoHH9OpZ9fuGje0vsm'; // Replace with your valid API key

// Function to fetch weather data
async function getWeatherData(lat, lon) {
    const url = `https://api.tomorrow.io/v4/timelines?location=${lat},${lon}&fields=temperature,humidity,windSpeed,precipitationIntensity&timesteps=1d&timezone=auto&apikey=${API_KEY}`;

    try {
        const response = await fetch(url);
        if (!response.ok) throw new Error(`Error: ${response.status}, ${response.statusText}`);
        const data = await response.json();
        return data.data.timelines[0].intervals;
    } catch (error) {
        console.error("Weather API error:", error);
    }
}

// Function to get water quantity based on soil moisture
function getWaterQuantity(stage, soilMoisture) {
    const baseWater = cropData.find(item => item.stage === stage)?.depth || 50;

    if (soilMoisture < 30) return baseWater + 10; // Increase by 10mm if soil is very dry
    if (soilMoisture >= 30 && soilMoisture < 50) return baseWater; // Standard water requirement
    return Math.max(baseWater - 10, 10); // Reduce water if soil is sufficiently moist
}

// Function to display irrigation results
function displayResults(waterQuantity, frequency, rainDates) {
    const resultDiv = document.getElementById("result");
    resultDiv.innerHTML = `
        <h2>Irrigation Recommendations</h2>
        <p>Water Quantity: Apply <strong>${waterQuantity} mm</strong> of water.</p>
        <p>Frequency: Irrigate every <strong>${frequency} days</strong>.</p>
        ${rainDates.length > 0 ? `<p>Rain expected on: ${rainDates.join(", ")}</p>` : '<p>No rain expected in the next few days.</p>'}
        <h3>Irrigation Efficiency Tips:</h3>
        <ul>
            <li>Use drip or furrow irrigation to minimize water loss.</li>
            <li>Water early in the morning or late in the evening to reduce evaporation.</li>
        </ul>
    `;
}

// Function to provide weather and irrigation recommendations
async function getWeatherAndRecommendations(lat, lon, crop, cropStage, soilMoisture) {
    const intervals = await getWeatherData(lat, lon);

    if (!intervals) return;

    let totalRainfall = 0;
    const rainDates = [];

    // Check the rainfall forecast for the next 5 days
    intervals.slice(0, 5).forEach(day => {
        const date = day.startTime.slice(0, 10);
        const rainAmount = day.values.precipitationIntensity || 0;
        totalRainfall += rainAmount;
        if (rainAmount > 0) rainDates.push(date);
    });

    // Irrigation timing recommendations
    if (totalRainfall >= 10) {
        console.log(`Rain expected on ${rainDates.join(", ")}. Skip irrigation to avoid over-watering.`);
    } else {
        console.log("Dry weather expected for the next 4-5 days.");
        if (soilMoisture < 40) {
            console.log("Soil moisture is low, irrigation needed within the next 24 hours.");
        } else {
            console.log("Soil moisture is adequate; monitor for changes.");
        }
    }

    // Water quantity recommendations
    const waterQuantity = getWaterQuantity(cropStage, soilMoisture);
    console.log(`Water Quantity: Apply ${waterQuantity} mm of water.`);

    // Frequency recommendations
    const frequency = cropData.find(item => item.crop === crop && item.stage === cropStage)?.frequency;
    console.log(`Frequency: Irrigate every ${frequency} days.`);

    // Display results
    displayResults(waterQuantity, frequency, rainDates);
}

// Form submission handler
document.getElementById("irrigationForm").addEventListener("submit", async function (event) {
    event.preventDefault();

    const lat = parseFloat(document.getElementById("latitude").value);
    const lon = parseFloat(document.getElementById("longitude").value);
    const crop = document.getElementById("cropType").value;
    const soilMoisture = parseFloat(document.getElementById("soilMoisture").value);
    const cropStage = document.getElementById("growthStage").value;

    await getWeatherAndRecommendations(lat, lon, crop, cropStage, soilMoisture);
});
